File: reference-python/shelfspoken/audible.py
"""shelfspoken.audible — candidate retrieval from Audible's public catalog API.

Unauthenticated, read-only. Returns plain Candidate dicts (see match.py).
This is the only network I/O in the matching path, kept separate so match()
stays pure.
"""

from __future__ import annotations

import json
import urllib.parse
import urllib.request

_API = "https://api.audible.com/1.0/catalog/products"
_GROUPS = "contributors,product_attrs,product_desc,series"


def search(title: str, author: str | None = None, num_results: int = 20, timeout: int = 20) -> list[dict]:
    params = {
        "title": title,
        "num_results": str(num_results),
        "response_groups": _GROUPS,
        "products_sort_by": "Relevance",
    }
    if author:
        params["author"] = author
    url = f"{_API}?{urllib.parse.urlencode(params)}"
    req = urllib.request.Request(url, headers={"User-Agent": "shelfspoken/0.1"})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        payload = json.load(resp)
    return [_to_candidate(p) for p in payload.get("products", [])]


def _to_candidate(p: dict) -> dict:
    return {
        "asin": p.get("asin"),
        "title": p.get("title") or "",
        "subtitle": p.get("subtitle"),
        "authors": [a.get("name", "") for a in p.get("authors") or []],
        "narrators": [n.get("name", "") for n in p.get("narrators") or []],
        "format_type": p.get("format_type"),
        "language": p.get("language"),
        "runtime_min": p.get("runtime_length_min"),
    }

File: src/audible.rs
//! Candidate retrieval from Audible's public catalog API (unauthenticated,
//! read-only). The only network I/O in the matching path.

use crate::matcher::Candidate;
use serde_json::Value;

const API: &str = "https://api.audible.com/1.0/catalog/products";
const GROUPS: &str = "contributors,product_attrs,product_desc,series";

pub fn search(title: &str, author: Option<&str>) -> Result<Vec<Candidate>, Box<dyn std::error::Error>> {
    let mut req = ureq::get(API)
        .query("title", title)
        .query("num_results", "20")
        .query("response_groups", GROUPS)
        .query("products_sort_by", "Relevance")
        .set("User-Agent", "narratarr/0.1");
    if let Some(a) = author {
        req = req.query("author", a);
    }
    let payload: Value = req.timeout(std::time::Duration::from_secs(20)).call()?.into_json()?;

    let products = payload["products"].as_array().cloned().unwrap_or_default();
    Ok(products.iter().map(to_candidate).collect())
}

fn names(v: &Value) -> Vec<String> {
    v.as_array()
        .map(|arr| {
            arr.iter()
                .filter_map(|p| p["name"].as_str().map(str::to_string))
                .collect()
        })
        .unwrap_or_default()
}

fn to_candidate(p: &Value) -> Candidate {
    Candidate {
        asin: p["asin"].as_str().map(str::to_string),
        title: p["title"].as_str().unwrap_or_default().to_string(),
        subtitle: p["subtitle"].as_str().map(str::to_string),
        authors: names(&p["authors"]),
        narrators: names(&p["narrators"]),
        format_type: p["format_type"].as_str().map(str::to_string),
        language: p["language"].as_str().map(str::to_string),
        runtime_min: p["runtime_length_min"].as_i64(),
    }
}
